### PR TITLE
Catch WSODs and provide a means for recovery for end users

### PR DESCRIPTION
**Pull request by felixarntz**
This is the PR accompanying https://core.trac.wordpress.org/ticket/44458.

This is only for development, discussion should continue to happen on the Trac ticket as usual.

### CODE REVIEW

**Comment by felixarntz** [comment]
The initial commit here is based on the state of [44458.6.diff](https://core.trac.wordpress.org/attachment/ticket/44458/44458.6.diff).

File: src/wp-admin/plugins.php [hunk 1]
```diff
@@ -480,6 +501,8 @@
 		$errmsg = __( 'You cannot delete a plugin while it is active on the main site.' );
 	} elseif ( isset( $_GET['charsout'] ) ) {
 		$errmsg = sprintf( __( 'The plugin generated %d characters of <strong>unexpected output</strong> during activation. If you notice &#8220;headers already sent&#8221; messages, problems with syndication feeds or other issues, try deactivating or removing this plugin.' ), $_GET['charsout'] );
+	} elseif ( 'resuming' === $_GET['error'] ) {
+		$errmsg = __( 'Plugin could not be resumed because it tri§ggered a <strong>fatal error</strong>.' );
```
Comment (afragen) [inline src/wp-admin/plugins.php]: Typo in `triggered`

**Comment by felixarntz** [comment]
I'm not entirely sure about dbe5bc0bc631fe819902667ad8798e7de509b940 yet (see https://core.trac.wordpress.org/ticket/44458#comment:33), we might need to discuss this further.

File: src/wp-admin/includes/class-wp-plugins-list-table.php [hunk 2]
```diff
@@ -421,24 +428,35 @@ protected function get_views() {
 
 			switch ( $type ) {
 				case 'all':
+					/* translators: %s: plugin count */
```
Comment (jeremyfelt) [inline src/wp-admin/includes/class-wp-plugins-list-table.php]: Other than the new `paused` case, these translator comments should go in their own docs/l10n commit under another ticket.

File: src/wp-admin/includes/plugin.php [hunk 3]
```diff
@@ -959,6 +1049,59 @@ function delete_plugins( $plugins, $deprecated = '' ) {
...
+			add_query_arg(
+				'_error_nonce',
+				wp_create_nonce( 'plugin-resume-error_' . $plugin ),
+				$redirect
+			)
+		);
+
+		// Load the plugin to test whether it throws a fatal error.
+		ob_start();
+		$plugin_path = WP_PLUGIN_DIR . '/' . $plugin;
```
Comment (spacedmonkey) [inline src/wp-admin/includes/plugin.php]: Could you use `plugin_sandbox_scrape` here?

File: src/wp-includes/class-wp-paused-extensions-storage.php [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+		if ( ! empty( $extension ) ) {
+			return array(
+				'key'         => $this->meta_prefix . $extension,
+				'compare_key' => '=',
+			);
+		}
+
+		return array(
+			'key'         => $this->meta_prefix . '%',
+			'compare_key' => 'LIKE',
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-paused-extensions-storage.php]: Why is there a like here?

File: src/wp-includes/error-protection.php [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	if ( empty( $callback ) || empty( $extension ) ) {
+		return false;
+	}
+
+	// Handle manually since the regular APIs do not expose this functionality.
+	if ( $network_wide ) {
+		$site_meta_query_clause = call_user_func( $callback )->get_site_meta_query_clause( $extension );
+		return delete_metadata( 'blog', 0, $site_meta_query_clause['key'], '', true );
```
Comment (spacedmonkey) [inline src/wp-includes/error-protection.php:120]: Check if site meta is supported here.

File: src/wp-includes/class-wp-paused-extensions-storage.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-paused-extensions-storage.php]: This clause is only for cases where you'd want to query sites that have _any_ paused extensions. This isn't currently used, all queries in use so far query sites by specific extensions paused.

File: src/wp-admin/includes/plugin.php [hunk 6]
```diff
@@ -2066,3 +2207,33 @@ function wp_add_privacy_policy_content( $plugin_name, $policy_text ) {
...
+
+	if ( ! current_user_can( 'deactivate_plugins' ) ) {
+		return;
+	}
+
+	if ( ! isset( $GLOBALS['_paused_plugins'] ) || empty( $GLOBALS['_paused_plugins'] ) ) {
+		return;
+	}
+
+	echo sprintf(
```
Comment (spacedmonkey) [inline src/wp-admin/includes/plugin.php]: `echo sprintf(` === `printf(`

File: src/wp-includes/error-protection.php (hunk 5 above)
Comment (spacedmonkey) [reply src/wp-includes/error-protection.php:120]: We could add a `delete_site_meta_by_key` similar to [delete_post_meta_by_key](https://developer.wordpress.org/reference/functions/delete_post_meta_by_key/)

File: src/wp-admin/includes/class-wp-plugins-list-table.php [hunk 7]
```diff
@@ -833,12 +873,54 @@ public function single_row( $item ) {
...
+						$notice_text = __( 'This plugin failed to load properly and was paused within the admin backend.' );
+						if ( $screen->in_admin( 'network' ) && $paused_on_network_sites_count ) {
+							$notice_text = sprintf(
+								/* translators: %s: number of sites */
+								_n( 'This plugin failed to load properly and was paused within the admin backend for %s site.', 'This plugin failed to load properly and was paused within the admin backend for %s sites.', $paused_on_network_sites_count ),
+								number_format_i18n( $paused_on_network_sites_count )
+							);
+						}
+
+						echo sprintf( '<p><span class="dashicons dashicons-warning"></span> <strong>%s</strong></p>', $notice_text );
```
Comment (afragen) [inline src/wp-admin/includes/class-wp-plugins-list-table.php]: Another `echo sprintf` === `printf`

File: src/wp-includes/error-protection.php [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+		if ( is_readable( $php_error_pluggable ) ) {
+			/*
+			 * This drop-in should control the HTTP status code and print the
+			 * HTML markup indicating that a PHP error occurred. Alternatively,
+			 * `wp_die()` can be used.
+			 */
+			require_once $php_error_pluggable;
+			die();
+		}
+
```
Comment (spacedmonkey) [inline src/wp-includes/error-protection.php]: It is likely we will need to add `wp_load_translations_early();` here

File: src/wp-includes/load.php [hunk 9]
```diff
@@ -1154,6 +1262,98 @@ function wp_doing_ajax() {
...
+		'edit-theme-plugin-file', // Saving changes in the core code editor.
+		'heartbeat',              // Keep the heart beating.
+		'install-plugin',         // Installing a new plugin.
+		'install-theme',          // Installing a new theme.
+		'search-plugins',         // Searching in the list of plugins.
+		'search-install-plugins', // Searching for a plugin in the plugin install screen.
+		'update-plugin',          // Update an existing plugin.
+		'update-theme',           // Update an existing theme.
+	);
+
```
Comment (spacedmonkey) [inline src/wp-includes/load.php:1355]: Add filter here.

Comment (felixarntz) [reply src/wp-includes/load.php:1355]: We can consider whether this makes sense in a future release, definitely in separate ticket. If necessary, the `is_protected_endpoint` filter can already be used for it.

File: src/wp-includes/error-protection.php (hunk 8 above)
Comment (felixarntz) [reply src/wp-includes/error-protection.php]: Good point. I'll defer this to @schlessera - we either need to ensure those functions are loaded, or (worse) have alternate clauses where one variant does not use `__()` but instead the untranslated string directly.

File: src/wp-includes/error-protection.php (hunk 5 above)
Comment (felixarntz) [reply src/wp-includes/error-protection.php:120]: Makes sense, but I think that can happen in a separate ticket later.

File: src/wp-admin/includes/class-wp-plugins-list-table.php [hunk 10]
```diff
@@ -833,12 +873,54 @@ public function single_row( $item ) {
 					 * @param array    $plugin_data An array of plugin data.
 					 * @param string   $status      Status of the plugin. Defaults are 'All', 'Active',
 					 *                              'Inactive', 'Recently Activated', 'Upgrade', 'Must-Use',
-					 *                              'Drop-ins', 'Search'.
+					 *                              'Drop-ins', 'Search', 'Paused'
```
Comment (afragen) [inline src/wp-admin/includes/class-wp-plugins-list-table.php]: Need a period at the end?

File: src/wp-admin/includes/class-wp-plugins-list-table.php [hunk 11]
```diff
@@ -887,7 +969,7 @@ public function single_row( $item ) {
 		 * @param array  $plugin_data An array of plugin data.
 		 * @param string $status      Status of the plugin. Defaults are 'All', 'Active',
 		 *                            'Inactive', 'Recently Activated', 'Upgrade', 'Must-Use',
-		 *                            'Drop-ins', 'Search'.
+		 *                            'Drop-ins', 'Search', 'Paused'
```
Comment (afragen) [inline src/wp-admin/includes/class-wp-plugins-list-table.php]: Another period?

File: src/wp-includes/error-protection.php (hunk 8 above)
Comment (schlessera) [reply src/wp-includes/error-protection.php]: We can just provide a stub in that case.

File: src/wp-includes/load.php (hunk 9 above)
Comment (schlessera) [reply src/wp-includes/load.php:1355]: I think it makes sense to add the filter here from the start. AJAX actions are just as crucial for some admin functionality than regular endpoints.

File: src/wp-includes/class-wp-shutdown-handler.php [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+	/**
+	 * Returns the error message markup to display in the default error template.
+	 *
+	 * @since 5.1.0
+	 *
+	 * @return string Error message HTML output.
+	 */
+	protected function get_error_message_markup() {
+		if ( ! function_exists( '__' ) ) {
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-shutdown-handler.php:143]: Why not call `	wp_load_translations_early();` here?

Comment (schlessera) [reply src/wp-includes/class-wp-shutdown-handler.php:143]: Not sure, but that sounds like it might be too risky for a shutdown handler...

Comment (felixarntz) [reply src/wp-includes/class-wp-shutdown-handler.php:143]: I agree with @schlessera. We might not even have the database initialized and thus might not even know which locale to use. Stubbing `__()` is not risky in that regard.

Comment (spacedmonkey) [reply src/wp-includes/class-wp-shutdown-handler.php:143]: But that is the point of the function, it doesn’t call the database or cache, but does it best to load translations. It loads in wp_db class super early. This function is designed exactly for this...

Comment (spacedmonkey) [reply src/wp-includes/class-wp-shutdown-handler.php:143]: Stubbing __() is safe, but it also means that the translation do not work at all. Trying to support translations if possible.

**Comment by TimothyBJacobs** [comment]
I opened #4 as requested to implement an opt-out mechanism via a plugin header.

**Comment by felixarntz** [comment]
This has been fixed (see https://core.trac.wordpress.org/ticket/44458#comment:56). Further work should happen through separate Trac tickets and patches or PRs.

**Pull request closed without merging**